src/main/java/org/mskcc/cmo/metadb/service/impl/MessageHandlingServiceImpl.java

```diff
@@ -76,34 +79,56 @@ public void run() {
                                     RequestStatusLogger.StatusType.REQUEST_WITH_MISSING_SAMPLES);
                             continue;
                         }
+                        Map<String, List<SampleMetadata>> patientSamplesMap = new HashMap<>();
+
                         // generate cmo labels for each sample metadata
                         List<SampleMetadata> updatedSamples = new ArrayList<>();
-                        for (SampleMetadata sample : request.getSamples()) {
-                            // generate cmo label
-                            String sampleCmoLabel = cmoLabelGeneratorService.generateCmoSampleLabel(sample);
+                        for (SampleMetadata sample : samples) {
+                            // get or request existing patient samples and update patient sample mapping
+                            if (!patientSamplesMap.containsKey(sample.getCmoPatientId())) {
+                                Message reply = messagingGateway.request(PATIENT_SAMPLES_REQUEST_TOPIC,
+                                        sample.getCmoPatientId());
+                                String replyData = mapper.readValue(
+                                        new String(reply.getData(), StandardCharsets.UTF_8),
+                                        String.class);
+                                patientSamplesMap.put(sample.getCmoPatientId(),
+                                        Arrays.asList(mapper.convertValue(replyData,
+                                                SampleMetadata[].class)));
+                            }
+                            List<SampleMetadata> existingSamples =
+                                    patientSamplesMap.get(sample.getCmoPatientId());
+
+                            // TODO resolve any issues that arise with errors in generating cmo label
+                            String sampleCmoLabel = cmoLabelGeneratorService.generateCmoSampleLabel(sample,
+                                    existingSamples);
                             sample.setCmoSampleName(sampleCmoLabel);
                             updatedSamples.add(sample);
+                            // update patient sample map also
+                            existingSamples.add(sample);
+                            patientSamplesMap.put(sample.getCmoPatientId(), existingSamples);
                         }
                         // save the updated samples to the request and publish to the igo new request topic
-                        request.setSamples(updatedSamples);
-                        // publish to igo new request topic
-                        messagingGateway.publish(request.getRequestId(),
-                                    IGO_NEW_REQUEST_TOPIC,
-                                    mapper.writeValueAsString(request));
+                        Map<String, Object> requestJsonMap = mapper.readValue(requestJson, Map.class);
+                        requestJsonMap.put("samples", updatedSamples);
```

Comment:
Again, in an effort to cut down the size of this loop, maybe introduce a method that takes a list of SampleMetadata and publishes it.

Also do you need to introduce updatedSamples list?  Is there any reason you cannot use samples, the return from getSamplesFromRequestJson() ?